File: 2018/day03.php
<?php

class Fabric
{
    public $id;
    public $x;
    public $y;
    public $w;
    public $h;

    public function __construct(int $id, int $x, int $y, int $w, int $h)
    {
        $this->id = $id;
        $this->x = $x;
        $this->y = $y;
        $this->w = $w;
        $this->h = $h;
    }
}

$fh = fopen(__DIR__.'/inputs/03.txt', 'r');
//$fabrics = [];
// $material[x][y] = [ ids ]
$material = [];

$maxX = 0;
$maxY = 0;

while (!feof($fh)) {
    $line = trim(fgets($fh));
    if (empty($line)) {
        continue;
    }

    $matches = [];
    $matched = preg_match('/^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$/', $line, $matches);

    if ($matched !== 1) {
        continue;
    }

    $f = new Fabric(intval($matches[1]), intval($matches[2]), intval($matches[3]), intval($matches[4]), intval($matches[5]));
    $maxX = max($maxX, $f->x + $f->w);
    $maxY = max($maxY, $f->y + $f->h);

    for ($y = $f->y; $y < ($f->y + $f->h); $y++) {
        for ($x = $f->x; $x < ($f->x + $f->w); $x++) {
            if (!array_key_exists($x, $material)) {
                $material[$x] = [];
            }

            if (!array_key_exists($y, $material[$x])) {
                $material[$x][$y] = 0;
            }

            $material[$x][$y]++;
        }
    }
}

fclose($fh);

$coverage = 0;

for ($i = 0; $i <= $maxY; $i++) {
    for ($j = 0; $j <= $maxX; $j++) {
        if (array_key_exists($j, $material) && array_key_exists($i, $material[$j])) {
            if ($material[$j][$i] > 1) {
                $coverage++;
            }
        }
    }
}

/*$coverage = 0;
array_walk_recursive($material, function($item, $key) use (&$coverage) {
    if ($item > 1) {
        $coverage++;
    }
});*/

var_dump($coverage);

File: 2018/day03.py
import re
import collections

Fabric = collections.namedtuple('Fabric', [ 'id', 'x', 'y', 'w', 'h' ])

def parse(line):
    m = re.search('^#(\d+) @ (\d+),(\d+): (\d+)x(\d+)$', line)
    return Fabric(id=int(m.group(1)), x=int(m.group(2)), y=int(m.group(3)), w=int(m.group(4)), h=int(m.group(5)))

with open('inputs/03.txt', 'r') as f:
    lines = f.read().splitlines()
    fabrics = [parse(line) for line in lines]

    min_x = min([f.x for f in fabrics])
    min_y = min([f.y for f in fabrics])
    max_x = max([f.x + f.w for f in fabrics ])
    max_y = max([f.y + f.h for f in fabrics ])

    grid = [ [ 0 for i in range(0, max_x + 1) ] for i in range(0, max_y + 1) ]

    for f in fabrics:
        for y in range(f.y, f.y + f.h):
            for x in range(f.x, f.x + f.w):
                grid[x][y] += 1

    part_1 = sum([ sum([ 1 if count > 1 else 0 for count in inner ]) for inner in grid ])

    print(part_1, a)
